File: tools/lib/browser_launcher.rb
# SPDX-License-Identifier: MPL-2.0

require "fileutils"
require_relative "defines"
require_relative "dev_browser"
require_relative "utils"

module FelesBuild
  # browser を起こして、閉じるまで出るものを読む。
  module BrowserLauncher
    LOGGER = Utils::Logger.new("launcher")

    # Firefox の出力は色で読む(上から順に、当たったものの色)
    COLORS = {
      /MOZ_CRASH|JavaScript error:|console\.error|\] Errors|\[fluent\] (Couldn't find a message:|Missing)|EGL Error:/ => 31,
      /console\.warn|WARNING:|\[WARN|JavaScript warning:/ => 33,
      /console\.debug/ => 36,
    }.freeze

    def self.command(port)
      [Defines::BIN_PATH_EXE,
       "--profile", Defines::PATHS[:profile_test],
       "--remote-debugging-port", port.to_s,
       # dev だけ: BiDi から about:/chrome: のページの中も評価できるように(手元で drop の様子を見るため)
       "--remote-allow-system-access",
       "--jsdebugger"]
    end

    def self.run(port = 5180)
      # 素の Firefox には buildid2 の patch が無く、build が変わっても startup cache が
      # 古いまま残る(chrome registry がずれる)。開く前に捨てて読み直させる。
      FileUtils.rm_rf(File.join(Defines::PATHS[:profile_test], "startupCache")) if Defines::STOCK_FIREFOX

      # 同じ profile の先客が居ると、Firefox はそちらに渡して、こちらはすぐ終わる。
      # 黙って渡すと「起きてすぐ閉じた」ように見えるので、先に言う。
      DevBrowser.running.each do |pid, _|
        LOGGER.warn "すでに開発用の noraneko が動いている (pid #{pid})。"
        LOGGER.warn "この profile は先客に渡るので、いま建てたものは映らない。"
        LOGGER.warn "閉じてから起こすなら: deno task feles-build stop"
      end

      cmd = command(port)
      puts "[launcher] Launching browser with command: #{cmd.join(' ')}"
      Utils.run_with_logging(cmd) do |stream, raw|
        line = raw.chomp
        if stream == :stderr && line.match?(%r{^WebDriver BiDi listening on ws://})
          puts "nora-{bbd11c51-3be9-4676-b912-ca4c0bdcab94}-webdriver"
        end
        color = COLORS.find { |pattern, _| line.match?(pattern) }&.last
        puts color ? "\e[#{color}m#{line}\e[0m" : line
      end
      puts "[launcher] Browser Closed"
    end
  end
end

File: tools/src/browser_launcher.ts
// SPDX-License-Identifier: MPL-2.0

import { BIN_PATH_EXE, PATHS, STOCK_FIREFOX } from "./defines.ts";
import { ProcessUtils, safeRemove } from "./utils.ts";
import * as path from "@std/path";

/**
 * Minimal port of tools/lib/browser_launcher.rb
 */

function printFirefoxLog(line: string) {
  if (
    /MOZ_CRASH|JavaScript error:|console\.error|\] Errors|\[fluent\] Couldn't find a message:|\[fluent\] Missing|EGL Error:/.test(
      line,
    )
  ) {
    console.log(`\x1b[31m${line}\x1b[0m`);
  } else if (/console\.warn|WARNING:|\[WARN|JavaScript warning:/.test(line)) {
    console.log(`\x1b[33m${line}\x1b[0m`);
  } else if (/console\.debug/.test(line)) {
    console.log(`\x1b[36m${line}\x1b[0m`);
  } else {
    console.log(line);
  }
}

export async function browserCommand(port: number): Promise<string[]> {
  return [
    BIN_PATH_EXE,
    "--profile",
    PATHS.profile_test,
    "--remote-debugging-port",
    String(port),
    // dev だけ: BiDi から about:/chrome: のページの中も評価できるように(手元で drop の様子を見るため)
    "--remote-allow-system-access",
    "--wait-for-browser",
    "--jsdebugger",
  ];
}

export async function run(port = 5180): Promise<void> {
  // Stock Firefox lacks the noraneko-runtime nsAppRunner buildid2 patch that
  // invalidates the startup cache when the build changes, so the chrome
  // registry (omni.ja chrome.manifest) can go stale across runs. Drop the
  // profile's startup cache before launch to force a fresh read.
  if (STOCK_FIREFOX) {
    safeRemove(path.join(PATHS.profile_test, "startupCache"));
  }

  const cmd = await browserCommand(port);

  console.log("[launcher] Launching browser with command: " + cmd.join(" "));

  await ProcessUtils.runCommandWithLogging(
    cmd,
    (stream: "stdout" | "stderr", line: string) => {
      if (stream === "stderr") {
        const m = line.match(/^WebDriver BiDi listening on (ws:\/\/.*)/);
        if (m) {
          console.log("nora-{bbd11c51-3be9-4676-b912-ca4c0bdcab94}-webdriver");
        }
      }
      printFirefoxLog(line.trim());
    },
  );

  console.log("[launcher] Browser Closed");
}
